bVNC/src/main/java/com/undatech/opaque/ConnectionGridActivity.java

```diff
@@ -500,7 +510,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
                 IntroTextDialog.showIntroTextIfNecessary(this, database, true);
             } else {
                 togglingMasterPassword = true;
-                if (Utils.querySharedPreferenceBoolean(this, Constants.masterPasswordEnabledTag)) {
+                if (Utils.querySharedPreferenceBoolean(this, Constants.masterPasswordEnabledTag) || !((Utils.querySharedPreferenceString(this, Constants.masterPassword, "")).equals("")) && Utils.querySharedPreferenceBoolean(this, Constants.onlyLockConnectionEditing)) {
```

Comment:
More safely, this is written "".equals(Utils.querySharedPreferenceString(this, Constants.masterPassword, "")).

Also, please take this condition out to an appropriately named separate method because it has gotten quite complex.